{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "F7Cerjso6Ptx"
      },
      "outputs": [],
      "source": [
        "import os\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import tensorflow as tf\n",
        "from tensorflow.keras.preprocessing.image import ImageDataGenerator\n",
        "from tensorflow.keras.callbacks import ModelCheckpoint\n",
        "from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout, GlobalAveragePooling2D\n",
        "from tensorflow.keras.models import Sequential\n",
        "from tensorflow.keras.optimizers import Adam\n",
        "from PIL import Image, ImageFile\n",
        "ImageFile.LOAD_TRUNCATED_IMAGES = True"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "ljfUKROb6XRZ",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "3dff1342-01f4-4592-a480-8be77ebdb923"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n"
          ]
        }
      ],
      "source": [
        "# Mount Google Drive to access dataset\n",
        "from google.colab import drive\n",
        "drive.mount('/content/drive', force_remount=True)\n",
        "\n",
        "# Set up directory paths\n",
        "base_dir = '/content/drive/MyDrive/Dataset Gabungan II Crop + Background + Split'\n",
        "h5_dir = '/content/CNN_Model2_Gabungan.h5'\n",
        "train_dir = os.path.join(base_dir, 'training')\n",
        "test_dir = os.path.join(base_dir, 'validation')"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# # Load the pre-trained model from the .h5 file\n",
        "# model = tf.keras.models.load_model(h5_dir)"
      ],
      "metadata": {
        "id": "JbdBvS7_7Y8b"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "uZyJbpj56XZ4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "079f2390-aa83-4a4d-b4c3-21d2d63cdb58"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Found 6744 images belonging to 6 classes.\n",
            "Found 1690 images belonging to 6 classes.\n"
          ]
        }
      ],
      "source": [
        "# Set up data generators for training and validation\n",
        "train_datagen = ImageDataGenerator(rescale=1./255,\n",
        "                                   rotation_range=20,\n",
        "                                   brightness_range=(0.8,1.2),\n",
        "                                   horizontal_flip=True)\n",
        "\n",
        "val_datagen = ImageDataGenerator(rescale=1./255,\n",
        "                                 rotation_range=20,\n",
        "                                 brightness_range=(0.8,1.2),\n",
        "                                 horizontal_flip=True)\n",
        "\n",
        "train_generator = train_datagen.flow_from_directory(train_dir,\n",
        "                                                    target_size=(100, 100),\n",
        "                                                    batch_size=16,\n",
        "                                                    class_mode='categorical')\n",
        "\n",
        "val_generator = val_datagen.flow_from_directory(test_dir,\n",
        "                                                target_size=(100, 100),\n",
        "                                                batch_size=16,\n",
        "                                                class_mode='categorical')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "kY1tqlvU6XlD"
      },
      "outputs": [],
      "source": [
        "# Define the model\n",
        "model = Sequential()\n",
        "\n",
        "# Add convolutional layers\n",
        "model.add(Conv2D(32, (3, 3), padding='same', activation='relu', input_shape=(100, 100, 3)))\n",
        "model.add(Conv2D(32, (3, 3), padding='same', activation='relu'))\n",
        "model.add(MaxPooling2D((2, 2)))\n",
        "\n",
        "model.add(Conv2D(64, (3, 3), padding='same', activation='relu'))\n",
        "model.add(Conv2D(64, (3, 3), padding='same', activation='relu'))\n",
        "model.add(MaxPooling2D((2, 2)))\n",
        "\n",
        "model.add(Conv2D(128, (3, 3), padding='same', activation='relu'))\n",
        "model.add(Conv2D(128, (3, 3), padding='same', activation='relu'))\n",
        "model.add(MaxPooling2D((2, 2)))\n",
        "\n",
        "# Replace Flatten() with GlobalAveragePooling2D()\n",
        "model.add(GlobalAveragePooling2D())\n",
        "\n",
        "model.add(Dense(256, activation='relu'))\n",
        "model.add(Dropout(0.5))\n",
        "model.add(Dense(128, activation='relu'))\n",
        "model.add(Dropout(0.5))\n",
        "model.add(Dense(6, activation='softmax'))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "enVtHDG46ezi"
      },
      "outputs": [],
      "source": [
        "# Compile the model\n",
        "opt = Adam(learning_rate=0.0001)\n",
        "model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])\n",
        "\n",
        "# Set up checkpoint callback\n",
        "checkpoint = ModelCheckpoint(\"CNN_Model2_Gabungan.h5\",\n",
        "                             monitor='val_accuracy',\n",
        "                             verbose=1,\n",
        "                             save_best_only=True,\n",
        "                             mode='max')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "YC6WIoSF6hW9",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "f138290a-9c82-4594-accd-cf3d44ac2499"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Epoch 1/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.7429 - accuracy: 0.2338\n",
            "Epoch 1: val_accuracy improved from -inf to 0.23669, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 867s 2s/step - loss: 1.7429 - accuracy: 0.2338 - val_loss: 1.7185 - val_accuracy: 0.2367\n",
            "Epoch 2/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.7293 - accuracy: 0.2281\n",
            "Epoch 2: val_accuracy did not improve from 0.23669\n",
            "422/422 [==============================] - 553s 1s/step - loss: 1.7293 - accuracy: 0.2281 - val_loss: 1.7183 - val_accuracy: 0.2367\n",
            "Epoch 3/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.7284 - accuracy: 0.2269\n",
            "Epoch 3: val_accuracy did not improve from 0.23669\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.7284 - accuracy: 0.2269 - val_loss: 1.7256 - val_accuracy: 0.2367\n",
            "Epoch 4/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.7228 - accuracy: 0.2334\n",
            "Epoch 4: val_accuracy did not improve from 0.23669\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.7228 - accuracy: 0.2334 - val_loss: 1.7126 - val_accuracy: 0.2367\n",
            "Epoch 5/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.7108 - accuracy: 0.2476\n",
            "Epoch 5: val_accuracy improved from 0.23669 to 0.26923, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 546s 1s/step - loss: 1.7108 - accuracy: 0.2476 - val_loss: 1.6808 - val_accuracy: 0.2692\n",
            "Epoch 6/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6883 - accuracy: 0.2657\n",
            "Epoch 6: val_accuracy improved from 0.26923 to 0.27929, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 542s 1s/step - loss: 1.6883 - accuracy: 0.2657 - val_loss: 1.6681 - val_accuracy: 0.2793\n",
            "Epoch 7/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6778 - accuracy: 0.2792\n",
            "Epoch 7: val_accuracy improved from 0.27929 to 0.28107, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.6778 - accuracy: 0.2792 - val_loss: 1.6676 - val_accuracy: 0.2811\n",
            "Epoch 8/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6776 - accuracy: 0.2758\n",
            "Epoch 8: val_accuracy improved from 0.28107 to 0.28166, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 550s 1s/step - loss: 1.6776 - accuracy: 0.2758 - val_loss: 1.6608 - val_accuracy: 0.2817\n",
            "Epoch 9/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6713 - accuracy: 0.2826\n",
            "Epoch 9: val_accuracy improved from 0.28166 to 0.28639, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 543s 1s/step - loss: 1.6713 - accuracy: 0.2826 - val_loss: 1.6575 - val_accuracy: 0.2864\n",
            "Epoch 10/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6638 - accuracy: 0.2917\n",
            "Epoch 10: val_accuracy improved from 0.28639 to 0.30473, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 549s 1s/step - loss: 1.6638 - accuracy: 0.2917 - val_loss: 1.6492 - val_accuracy: 0.3047\n",
            "Epoch 11/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6561 - accuracy: 0.2982\n",
            "Epoch 11: val_accuracy did not improve from 0.30473\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.6561 - accuracy: 0.2982 - val_loss: 1.6347 - val_accuracy: 0.2982\n",
            "Epoch 12/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6442 - accuracy: 0.3141\n",
            "Epoch 12: val_accuracy improved from 0.30473 to 0.31716, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.6442 - accuracy: 0.3141 - val_loss: 1.6229 - val_accuracy: 0.3172\n",
            "Epoch 13/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6226 - accuracy: 0.3235\n",
            "Epoch 13: val_accuracy improved from 0.31716 to 0.32781, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.6226 - accuracy: 0.3235 - val_loss: 1.6087 - val_accuracy: 0.3278\n",
            "Epoch 14/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.6087 - accuracy: 0.3369\n",
            "Epoch 14: val_accuracy improved from 0.32781 to 0.33077, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.6087 - accuracy: 0.3369 - val_loss: 1.6037 - val_accuracy: 0.3308\n",
            "Epoch 15/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5909 - accuracy: 0.3455\n",
            "Epoch 15: val_accuracy improved from 0.33077 to 0.33195, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.5909 - accuracy: 0.3455 - val_loss: 1.6159 - val_accuracy: 0.3320\n",
            "Epoch 16/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5695 - accuracy: 0.3590\n",
            "Epoch 16: val_accuracy improved from 0.33195 to 0.34615, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.5695 - accuracy: 0.3590 - val_loss: 1.5707 - val_accuracy: 0.3462\n",
            "Epoch 17/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5566 - accuracy: 0.3588\n",
            "Epoch 17: val_accuracy improved from 0.34615 to 0.36095, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 543s 1s/step - loss: 1.5566 - accuracy: 0.3588 - val_loss: 1.5447 - val_accuracy: 0.3609\n",
            "Epoch 18/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5469 - accuracy: 0.3665\n",
            "Epoch 18: val_accuracy improved from 0.36095 to 0.36391, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 549s 1s/step - loss: 1.5469 - accuracy: 0.3665 - val_loss: 1.5358 - val_accuracy: 0.3639\n",
            "Epoch 19/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5346 - accuracy: 0.3725\n",
            "Epoch 19: val_accuracy improved from 0.36391 to 0.36923, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.5346 - accuracy: 0.3725 - val_loss: 1.5196 - val_accuracy: 0.3692\n",
            "Epoch 20/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5305 - accuracy: 0.3784\n",
            "Epoch 20: val_accuracy improved from 0.36923 to 0.37988, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.5305 - accuracy: 0.3784 - val_loss: 1.5253 - val_accuracy: 0.3799\n",
            "Epoch 21/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5201 - accuracy: 0.3820\n",
            "Epoch 21: val_accuracy did not improve from 0.37988\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.5201 - accuracy: 0.3820 - val_loss: 1.5077 - val_accuracy: 0.3746\n",
            "Epoch 22/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5168 - accuracy: 0.3916\n",
            "Epoch 22: val_accuracy did not improve from 0.37988\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.5168 - accuracy: 0.3916 - val_loss: 1.5013 - val_accuracy: 0.3645\n",
            "Epoch 23/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5115 - accuracy: 0.3866\n",
            "Epoch 23: val_accuracy improved from 0.37988 to 0.38047, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.5115 - accuracy: 0.3866 - val_loss: 1.4883 - val_accuracy: 0.3805\n",
            "Epoch 24/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.5080 - accuracy: 0.3855\n",
            "Epoch 24: val_accuracy did not improve from 0.38047\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.5080 - accuracy: 0.3855 - val_loss: 1.4860 - val_accuracy: 0.3787\n",
            "Epoch 25/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4971 - accuracy: 0.3846\n",
            "Epoch 25: val_accuracy improved from 0.38047 to 0.38462, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 543s 1s/step - loss: 1.4971 - accuracy: 0.3846 - val_loss: 1.4932 - val_accuracy: 0.3846\n",
            "Epoch 26/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4866 - accuracy: 0.3977\n",
            "Epoch 26: val_accuracy improved from 0.38462 to 0.40059, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 552s 1s/step - loss: 1.4866 - accuracy: 0.3977 - val_loss: 1.4873 - val_accuracy: 0.4006\n",
            "Epoch 27/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4810 - accuracy: 0.4001\n",
            "Epoch 27: val_accuracy did not improve from 0.40059\n",
            "422/422 [==============================] - 544s 1s/step - loss: 1.4810 - accuracy: 0.4001 - val_loss: 1.4927 - val_accuracy: 0.3828\n",
            "Epoch 28/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4757 - accuracy: 0.3980\n",
            "Epoch 28: val_accuracy improved from 0.40059 to 0.40947, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 543s 1s/step - loss: 1.4757 - accuracy: 0.3980 - val_loss: 1.4642 - val_accuracy: 0.4095\n",
            "Epoch 29/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4684 - accuracy: 0.4078\n",
            "Epoch 29: val_accuracy did not improve from 0.40947\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.4684 - accuracy: 0.4078 - val_loss: 1.4491 - val_accuracy: 0.3964\n",
            "Epoch 30/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4670 - accuracy: 0.4110\n",
            "Epoch 30: val_accuracy did not improve from 0.40947\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.4670 - accuracy: 0.4110 - val_loss: 1.4712 - val_accuracy: 0.4012\n",
            "Epoch 31/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4537 - accuracy: 0.4189\n",
            "Epoch 31: val_accuracy improved from 0.40947 to 0.43136, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.4537 - accuracy: 0.4189 - val_loss: 1.4451 - val_accuracy: 0.4314\n",
            "Epoch 32/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4436 - accuracy: 0.4198\n",
            "Epoch 32: val_accuracy did not improve from 0.43136\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.4436 - accuracy: 0.4198 - val_loss: 1.4344 - val_accuracy: 0.4136\n",
            "Epoch 33/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4387 - accuracy: 0.4207\n",
            "Epoch 33: val_accuracy did not improve from 0.43136\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.4387 - accuracy: 0.4207 - val_loss: 1.4701 - val_accuracy: 0.4095\n",
            "Epoch 34/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4357 - accuracy: 0.4282\n",
            "Epoch 34: val_accuracy did not improve from 0.43136\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.4357 - accuracy: 0.4282 - val_loss: 1.4308 - val_accuracy: 0.4296\n",
            "Epoch 35/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4314 - accuracy: 0.4278\n",
            "Epoch 35: val_accuracy did not improve from 0.43136\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.4314 - accuracy: 0.4278 - val_loss: 1.4257 - val_accuracy: 0.4290\n",
            "Epoch 36/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4266 - accuracy: 0.4269\n",
            "Epoch 36: val_accuracy did not improve from 0.43136\n",
            "422/422 [==============================] - 548s 1s/step - loss: 1.4266 - accuracy: 0.4269 - val_loss: 1.4315 - val_accuracy: 0.4296\n",
            "Epoch 37/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4209 - accuracy: 0.4388\n",
            "Epoch 37: val_accuracy improved from 0.43136 to 0.44793, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 547s 1s/step - loss: 1.4209 - accuracy: 0.4388 - val_loss: 1.4111 - val_accuracy: 0.4479\n",
            "Epoch 38/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4218 - accuracy: 0.4303\n",
            "Epoch 38: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 536s 1s/step - loss: 1.4218 - accuracy: 0.4303 - val_loss: 1.4255 - val_accuracy: 0.4314\n",
            "Epoch 39/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.4163 - accuracy: 0.4386\n",
            "Epoch 39: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 530s 1s/step - loss: 1.4163 - accuracy: 0.4386 - val_loss: 1.4444 - val_accuracy: 0.4266\n",
            "Epoch 40/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3990 - accuracy: 0.4399\n",
            "Epoch 40: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 529s 1s/step - loss: 1.3990 - accuracy: 0.4399 - val_loss: 1.4120 - val_accuracy: 0.4337\n",
            "Epoch 41/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3948 - accuracy: 0.4435\n",
            "Epoch 41: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 528s 1s/step - loss: 1.3948 - accuracy: 0.4435 - val_loss: 1.3991 - val_accuracy: 0.4462\n",
            "Epoch 42/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3905 - accuracy: 0.4518\n",
            "Epoch 42: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 533s 1s/step - loss: 1.3905 - accuracy: 0.4518 - val_loss: 1.3986 - val_accuracy: 0.4467\n",
            "Epoch 43/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3806 - accuracy: 0.4527\n",
            "Epoch 43: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 530s 1s/step - loss: 1.3806 - accuracy: 0.4527 - val_loss: 1.3986 - val_accuracy: 0.4473\n",
            "Epoch 44/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3779 - accuracy: 0.4474\n",
            "Epoch 44: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 528s 1s/step - loss: 1.3779 - accuracy: 0.4474 - val_loss: 1.4161 - val_accuracy: 0.4420\n",
            "Epoch 45/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3802 - accuracy: 0.4481\n",
            "Epoch 45: val_accuracy did not improve from 0.44793\n",
            "422/422 [==============================] - 534s 1s/step - loss: 1.3802 - accuracy: 0.4481 - val_loss: 1.4065 - val_accuracy: 0.4361\n",
            "Epoch 46/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3774 - accuracy: 0.4540\n",
            "Epoch 46: val_accuracy improved from 0.44793 to 0.45325, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 535s 1s/step - loss: 1.3774 - accuracy: 0.4540 - val_loss: 1.3850 - val_accuracy: 0.4533\n",
            "Epoch 47/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3659 - accuracy: 0.4591\n",
            "Epoch 47: val_accuracy did not improve from 0.45325\n",
            "422/422 [==============================] - 531s 1s/step - loss: 1.3659 - accuracy: 0.4591 - val_loss: 1.3929 - val_accuracy: 0.4432\n",
            "Epoch 48/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3704 - accuracy: 0.4524\n",
            "Epoch 48: val_accuracy did not improve from 0.45325\n",
            "422/422 [==============================] - 533s 1s/step - loss: 1.3704 - accuracy: 0.4524 - val_loss: 1.3814 - val_accuracy: 0.4420\n",
            "Epoch 49/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3586 - accuracy: 0.4659\n",
            "Epoch 49: val_accuracy improved from 0.45325 to 0.46391, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 529s 1s/step - loss: 1.3586 - accuracy: 0.4659 - val_loss: 1.3666 - val_accuracy: 0.4639\n",
            "Epoch 50/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3569 - accuracy: 0.4613\n",
            "Epoch 50: val_accuracy did not improve from 0.46391\n",
            "422/422 [==============================] - 535s 1s/step - loss: 1.3569 - accuracy: 0.4613 - val_loss: 1.3618 - val_accuracy: 0.4639\n",
            "Epoch 51/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3514 - accuracy: 0.4585\n",
            "Epoch 51: val_accuracy did not improve from 0.46391\n",
            "422/422 [==============================] - 537s 1s/step - loss: 1.3514 - accuracy: 0.4585 - val_loss: 1.3752 - val_accuracy: 0.4568\n",
            "Epoch 52/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3430 - accuracy: 0.4705\n",
            "Epoch 52: val_accuracy did not improve from 0.46391\n",
            "422/422 [==============================] - 534s 1s/step - loss: 1.3430 - accuracy: 0.4705 - val_loss: 1.3849 - val_accuracy: 0.4402\n",
            "Epoch 53/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3311 - accuracy: 0.4785\n",
            "Epoch 53: val_accuracy did not improve from 0.46391\n",
            "422/422 [==============================] - 530s 1s/step - loss: 1.3311 - accuracy: 0.4785 - val_loss: 1.3646 - val_accuracy: 0.4639\n",
            "Epoch 54/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3350 - accuracy: 0.4810\n",
            "Epoch 54: val_accuracy did not improve from 0.46391\n",
            "422/422 [==============================] - 529s 1s/step - loss: 1.3350 - accuracy: 0.4810 - val_loss: 1.3588 - val_accuracy: 0.4604\n",
            "Epoch 55/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3380 - accuracy: 0.4751\n",
            "Epoch 55: val_accuracy improved from 0.46391 to 0.47633, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 534s 1s/step - loss: 1.3380 - accuracy: 0.4751 - val_loss: 1.3313 - val_accuracy: 0.4763\n",
            "Epoch 56/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3308 - accuracy: 0.4730\n",
            "Epoch 56: val_accuracy did not improve from 0.47633\n",
            "422/422 [==============================] - 533s 1s/step - loss: 1.3308 - accuracy: 0.4730 - val_loss: 1.3571 - val_accuracy: 0.4574\n",
            "Epoch 57/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3160 - accuracy: 0.4880\n",
            "Epoch 57: val_accuracy improved from 0.47633 to 0.47929, saving model to CNN_Model2_Gabungan.h5\n",
            "422/422 [==============================] - 529s 1s/step - loss: 1.3160 - accuracy: 0.4880 - val_loss: 1.3465 - val_accuracy: 0.4793\n",
            "Epoch 58/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3273 - accuracy: 0.4751\n",
            "Epoch 58: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 539s 1s/step - loss: 1.3273 - accuracy: 0.4751 - val_loss: 1.3543 - val_accuracy: 0.4574\n",
            "Epoch 59/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3181 - accuracy: 0.4806\n",
            "Epoch 59: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 533s 1s/step - loss: 1.3181 - accuracy: 0.4806 - val_loss: 1.3386 - val_accuracy: 0.4722\n",
            "Epoch 60/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3179 - accuracy: 0.4878\n",
            "Epoch 60: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 534s 1s/step - loss: 1.3179 - accuracy: 0.4878 - val_loss: 1.3649 - val_accuracy: 0.4604\n",
            "Epoch 61/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3012 - accuracy: 0.4905\n",
            "Epoch 61: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 537s 1s/step - loss: 1.3012 - accuracy: 0.4905 - val_loss: 1.3166 - val_accuracy: 0.4746\n",
            "Epoch 62/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.3060 - accuracy: 0.4821\n",
            "Epoch 62: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 533s 1s/step - loss: 1.3060 - accuracy: 0.4821 - val_loss: 1.3369 - val_accuracy: 0.4657\n",
            "Epoch 63/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.2908 - accuracy: 0.4950\n",
            "Epoch 63: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 537s 1s/step - loss: 1.2908 - accuracy: 0.4950 - val_loss: 1.3473 - val_accuracy: 0.4787\n",
            "Epoch 64/100\n",
            "422/422 [==============================] - ETA: 0s - loss: 1.2888 - accuracy: 0.4936\n",
            "Epoch 64: val_accuracy did not improve from 0.47929\n",
            "422/422 [==============================] - 532s 1s/step - loss: 1.2888 - accuracy: 0.4936 - val_loss: 1.3409 - val_accuracy: 0.4657\n",
            "Epoch 65/100\n",
            "238/422 [===============>..............] - ETA: 3:34 - loss: 1.2832 - accuracy: 0.4900"
          ]
        },
        {
          "output_type": "error",
          "ename": "UnknownError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mUnknownError\u001b[0m                              Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-23-bcd58a81e5cd>\u001b[0m in \u001b[0;36m<cell line: 2>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Train the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m history = model.fit(train_generator,\n\u001b[0m\u001b[1;32m      3\u001b[0m                     \u001b[0mepochs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m                     \u001b[0mvalidation_data\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mval_generator\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m                     callbacks=[checkpoint])\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/keras/utils/traceback_utils.py\u001b[0m in \u001b[0;36merror_handler\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m     68\u001b[0m             \u001b[0;31m# To get the full stack trace, call:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     69\u001b[0m             \u001b[0;31m# `tf.debugging.disable_traceback_filtering()`\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 70\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwith_traceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfiltered_tb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     71\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     72\u001b[0m             \u001b[0;32mdel\u001b[0m \u001b[0mfiltered_tb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/tensorflow/python/eager/execute.py\u001b[0m in \u001b[0;36mquick_execute\u001b[0;34m(op_name, num_outputs, inputs, attrs, ctx, name)\u001b[0m\n\u001b[1;32m     50\u001b[0m   \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     51\u001b[0m     \u001b[0mctx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mensure_initialized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 52\u001b[0;31m     tensors = pywrap_tfe.TFE_Py_Execute(ctx._handle, device_name, op_name,\n\u001b[0m\u001b[1;32m     53\u001b[0m                                         inputs, attrs, num_outputs)\n\u001b[1;32m     54\u001b[0m   \u001b[0;32mexcept\u001b[0m \u001b[0mcore\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_NotOkStatusException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mUnknownError\u001b[0m: Graph execution error:\n\nFileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Dataset Gabungan II Crop + Background + Split/training/square/square (424).jpg'\nTraceback (most recent call last):\n\n  File \"/usr/local/lib/python3.10/dist-packages/tensorflow/python/ops/script_ops.py\", line 267, in __call__\n    ret = func(*args)\n\n  File \"/usr/local/lib/python3.10/dist-packages/tensorflow/python/autograph/impl/api.py\", line 642, in wrapper\n    return func(*args, **kwargs)\n\n  File \"/usr/local/lib/python3.10/dist-packages/tensorflow/python/data/ops/from_generator_op.py\", line 198, in generator_py_func\n    values = next(generator_state.get_iterator(iterator_id))\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/engine/data_adapter.py\", line 902, in wrapped_generator\n    for data in generator_fn():\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/engine/data_adapter.py\", line 1049, in generator_fn\n    yield x[i]\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/preprocessing/image.py\", line 116, in __getitem__\n    return self._get_batches_of_transformed_samples(index_array)\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/preprocessing/image.py\", line 370, in _get_batches_of_transformed_samples\n    img = image_utils.load_img(\n\n  File \"/usr/local/lib/python3.10/dist-packages/keras/utils/image_utils.py\", line 422, in load_img\n    with open(path, \"rb\") as f:\n\nFileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Dataset Gabungan II Crop + Background + Split/training/square/square (424).jpg'\n\n\n\t [[{{node PyFunc}}]]\n\t [[IteratorGetNext]] [Op:__inference_train_function_9703]"
          ]
        }
      ],
      "source": [
        "# Train the model\n",
        "history = model.fit(train_generator,\n",
        "                    epochs=100,\n",
        "                    validation_data=val_generator,\n",
        "                    callbacks=[checkpoint])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "3m41Ti3W6i5S"
      },
      "outputs": [],
      "source": [
        "# Re-evaluate the model\n",
        "loss, acc = model.evaluate(val_generator, verbose=2)\n",
        "print(\"Accuracy: {:5.2f}%\".format(100 * acc))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4ZSQbOAi6svy"
      },
      "outputs": [],
      "source": [
        "acc = history.history['accuracy']\n",
        "val_acc = history.history['val_accuracy']\n",
        "loss = history.history['loss']\n",
        "val_loss = history.history['val_loss']\n",
        "\n",
        "epochs = range(len(acc))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "UE7W0xtt6s6g"
      },
      "outputs": [],
      "source": [
        "# Create a figure with two subplots\n",
        "fig, ax = plt.subplots(nrows=1, ncols=2, figsize=(12, 6))\n",
        "\n",
        "# Plotting accuracy\n",
        "ax[0].plot(epochs, acc, 'r', linewidth=2, label='Training Accuracy')\n",
        "ax[0].plot(epochs, val_acc, 'b', linewidth=2, label='Validation Accuracy')\n",
        "ax[0].set_title('Training and Validation Accuracy', fontsize=16)\n",
        "ax[0].set_xlabel('Epoch', fontsize=14)\n",
        "ax[0].set_ylabel('Accuracy', fontsize=14)\n",
        "ax[0].legend(loc='lower right', fontsize=12)\n",
        "ax[0].grid()\n",
        "ax[0].tick_params(axis='both', labelsize=12)\n",
        "\n",
        "# Plotting loss\n",
        "ax[1].plot(epochs, loss, 'r', linewidth=2, label='Training Loss')\n",
        "ax[1].plot(epochs, val_loss, 'b', linewidth=2, label='Validation Loss')\n",
        "ax[1].set_title('Training and Validation Loss', fontsize=16)\n",
        "ax[1].set_xlabel('Epoch', fontsize=14)\n",
        "ax[1].set_ylabel('Loss', fontsize=14)\n",
        "ax[1].legend(loc='upper right', fontsize=12)\n",
        "ax[1].grid()\n",
        "ax[1].tick_params(axis='both', labelsize=12)\n",
        "\n",
        "plt.show()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "vu_WAmyX6w1t"
      },
      "outputs": [],
      "source": [
        "# Get the predicted labels for all images in the cross validation set\n",
        "pred = model.predict(val_generator)\n",
        "predicted_labels = np.argmax(pred, axis=1)\n",
        "\n",
        "# Get the true labels for all images in the cross validation set\n",
        "true_labels = val_generator.classes\n",
        "\n",
        "# Get the filenames of the images in the cross validation set\n",
        "filenames = val_generator.filenames\n",
        "\n",
        "# Get the indices of the misclassified images\n",
        "misclassified_indices = np.where(predicted_labels != true_labels)[0]\n",
        "\n",
        "# Get the misclassified images and their true and predicted labels\n",
        "misclassified_images = []\n",
        "misclassified_true_labels = []\n",
        "misclassified_predicted_labels = []\n",
        "\n",
        "for i in misclassified_indices:\n",
        "    misclassified_images.append(filenames[i])\n",
        "    misclassified_true_labels.append(true_labels[i])\n",
        "    misclassified_predicted_labels.append(predicted_labels[i])\n",
        "\n",
        "# Print the misclassified images and their true and predicted labels\n",
        "print(\"Misclassified Images:\")\n",
        "for i in range(len(misclassified_images)):\n",
        "    print(\"Image:\", misclassified_images[i], \"True Label:\", misclassified_true_labels[i], \"Predicted Label:\", misclassified_predicted_labels[i])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "wqN-rnLL6tGQ"
      },
      "outputs": [],
      "source": [
        "# Display some of the misclassified images\n",
        "num_images_to_display = 18\n",
        "fig, axes = plt.subplots(6, 3, figsize=(10, 10))\n",
        "axes = axes.ravel()\n",
        "\n",
        "# Map class indices to class names\n",
        "class_names = {v: k for k, v in val_generator.class_indices.items()}\n",
        "\n",
        "for i in range(num_images_to_display):\n",
        "    img = Image.open(os.path.join(test_dir, misclassified_images[i]))\n",
        "    true_label = class_names[misclassified_true_labels[i]]\n",
        "    predicted_label = class_names[misclassified_predicted_labels[i]]\n",
        "    axes[i].imshow(img)\n",
        "    axes[i].set_title(\"True Label: {}\\nPredicted Label: {}\".format(true_label, predicted_label))\n",
        "    axes[i].axis('off')\n",
        "plt.tight_layout()\n",
        "plt.show()"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}